cpu/sam0_common/periph/i2c.c

```diff
@@ -45,365 +47,208 @@
+                          (i2c_config[dev].flags & I2C_FLAG_RUN_STANDBY ?
+                              SERCOM_I2CM_CTRLA_RUNSTDBY : 0);
 
     /* Enable Smart Mode (ACK is sent when DATA.DATA is read) */
-    I2CSercom->CTRLB.reg = SERCOM_I2CM_CTRLB_SMEN;
+    bus(dev)->CTRLB.reg = SERCOM_I2CM_CTRLB_SMEN;
 
     /* Find and set baudrate. Read speed configuration. Set transfer
      * speed: SERCOM_I2CM_CTRLA_SPEED(0): Standard-mode (Sm) up to 100
      * kHz and Fast-mode (Fm) up to 400 kHz */
-    switch (speed) {
+    switch (i2c_config[dev].speed) {
         case I2C_SPEED_NORMAL:
-            tmp_baud = (int32_t)(((clock_source_speed + (2 * (100000)) - 1) / (2 * (100000))) - 5);
-            if (tmp_baud < 255 && tmp_baud > 0) {
-                I2CSercom->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(0);
-                I2CSercom->BAUD.reg = SERCOM_I2CM_BAUD_BAUD(tmp_baud);
-            }
-            break;
         case I2C_SPEED_FAST:
-            tmp_baud = (int32_t)(((clock_source_speed + (2 * (400000)) - 1) / (2 * (400000))) - 5);
-            if (tmp_baud < 255 && tmp_baud > 0) {
-                I2CSercom->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(0);
-                I2CSercom->BAUD.reg = SERCOM_I2CM_BAUD_BAUD(tmp_baud);
-            }
+            bus(dev)->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(0);
             break;
         case I2C_SPEED_HIGH:
-            tmp_baud = (int32_t)(((clock_source_speed + (2 * (3400000)) - 1) / (2 * (3400000))) - 1);
-            if (tmp_baud < 255 && tmp_baud > 0) {
-                I2CSercom->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(2);
-                I2CSercom->BAUD.reg = SERCOM_I2CM_BAUD_HSBAUD(tmp_baud);
-            }
+            bus(dev)->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(2);
             break;
         default:
             DEBUG("BAD BAUDRATE\n");
             return -2;
     }
-
+    /* Get the baudrate */
+    tmp_baud = (int32_t)(((CLOCK_CORECLOCK +
+               (2 * (i2c_config[dev].speed)) - 1) /
+               (2 * (i2c_config[dev].speed))) -
+               (i2c_config[dev].speed == I2C_SPEED_HIGH ? 1 : 5));
+    /* Ensure baudrate is within limits */
+    if (tmp_baud < 255 && tmp_baud > 0) {
+        bus(dev)->BAUD.reg = SERCOM_I2CM_BAUD_BAUD(tmp_baud);
+    }
     /* ENABLE I2C MASTER */
-    i2c_poweron(dev);
+    _i2c_poweron(dev);
 
     /* Start timeout if bus state is unknown. */
-    while ((I2CSercom->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN) {
+    while ((bus(dev)->STATUS.reg &
+          SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN) {
         if (timeout_counter++ >= SAMD21_I2C_TIMEOUT) {
             /* Timeout, force bus state to idle. */
-            I2CSercom->STATUS.reg = BUSSTATE_IDLE;
+            bus(dev)->STATUS.reg = BUSSTATE_IDLE;
         }
     }
     return 0;
 }
 
 int i2c_acquire(i2c_t dev)
 {
-    if (dev >= I2C_NUMOF) {
-        return -1;
-    }
+    assert(dev < I2C_NUMOF);
     mutex_lock(&locks[dev]);
```

Comment:
I like the assert, I assume once everything is ported we can either remove the int return (because all failures would be asserted) or at least update the comments in the i2c.h driver.

Should the mutex_lock be used or the mutex_trylock or should there be an option?

Reply:
@aabadie what's your opinion about assert vs return err_code ?

Reply:
+1 for the assert. Brief comparing with spi API, the spi_init returns void, I think it should be the same for i2c (sorry for noticing this only now).